libraries/abstractions/pkcs11/psa/iot_pkcs11_psa.c

```diff
@@ -0,0 +1,3476 @@
+		( NULL == pulObjectCount ) )
+	{
+		xResult = CKR_ARGUMENTS_BAD;
+		xDone = pdTRUE;
+	}
+
+	if( xResult == CKR_OK )
+	{
+		if( pxSession->pxFindObjectLabel == NULL )
+		{
+			xResult = CKR_OPERATION_NOT_INITIALIZED;
+			xDone = pdTRUE;
+		}
+
+		if( 0u == ulMaxObjectCount )
+		{
+			xResult = CKR_ARGUMENTS_BAD;
+			xDone = pdTRUE;
+		}
+
+		if( 1u != ulMaxObjectCount )
+		{
+			PKCS11_WARNING_PRINT( ( "WARN: Searching for more than 1 object not supported. \r\n" ) );
+		}
+
+		if( ( pdFALSE == xDone ) && ( ( CK_BBOOL ) CK_TRUE == pxSession->xFindObjectComplete ) )
+		{
+			*pulObjectCount = 0;
+			xResult = CKR_OK;
+			xDone = pdTRUE;
+		}
+	}
+
+	/* TODO: Re-inspect this previous logic. */
+	if( ( pdFALSE == xDone ) )
+	{
+		/* Try to find the object in module's list first. */
+		prvFindObjectInListByLabel( pxSession->pxFindObjectLabel, strlen( ( const char * ) pxSession->pxFindObjectLabel ), &xPalHandle, pxObject );
+
+		/* Check with the PAL if the object was previously stored. */
+		if( *pxObject != CK_INVALID_HANDLE )
+		{
+			*pulObjectCount = 1;
+		}
+		else
+		{
+
+			/* Check with the PSA if the object was previously stored. */
+			xPalHandle = PKCS11_PSA_FindObject( pxSession->pxFindObjectLabel, ( uint8_t ) strlen( ( const char * ) pxSession->pxFindObjectLabel ) );
+			if( xPalHandle != CK_INVALID_HANDLE )
+			{
+				xResult = prvAddObjectToList( xPalHandle, pxObject, pxSession->pxFindObjectLabel, strlen( ( const char * ) pxSession->pxFindObjectLabel ) );
+				*pulObjectCount = 1;
+			}
+			else
+			{
+				/* According to the PKCS #11 standard, not finding an object results in a CKR_OK return value with an object count of 0. */
+				*pulObjectCount = 0;
+				PKCS11_WARNING_PRINT( ( "WARN: Object with label '%s' not found. \r\n", ( char * ) pxSession->pxFindObjectLabel ) );
+			}
+		}
+	}
+
+	return xResult;
+}
+
+/**
+ * @brief Completes an object search operation.
+ *
+ * \sa C_FindObjectsInit(), C_FindObjects() which must be called before
+ * calling C_FindObjectsFinal().
+ *
+ * \note FindObjects parameters are shared by a session.  Calling
+ * C_FindObjectsInit(), C_FindObjects(), and C_FindObjectsFinal() with the
+ * same session across different tasks may lead to unexpected results.
+ *
+ *
+ * @param[in] xSession					  Handle of a valid PKCS #11 session.
+ *
+ * @return CKR_OK if successful.
+ * Else, see <a href="https://tiny.amazon.com/wtscrttv">PKCS #11 specification</a>
+ * for more information.
+ */
+CK_DECLARE_FUNCTION( CK_RV, C_FindObjectsFinal )( CK_SESSION_HANDLE xSession )
+{ /*lint !e9072 It's OK to have different parameter name. */
+	CK_RV xResult = PKCS11_SESSION_VALID_AND_MODULE_INITIALIZED( xSession );
+
+	P11SessionPtr_t pxSession = prvSessionPointerFromHandle( xSession );
+
+	/*
+	 * Check parameters.
+	 */
+	if( xResult == CKR_OK )
+	{
+		if( pxSession->pxFindObjectLabel == NULL )
+		{
+			xResult = CKR_OPERATION_NOT_INITIALIZED;
+		}
+	}
+
+	if( xResult == CKR_OK )
+	{
+		/*
+		 * Clea
```

Comment:
This lock might be redundant in the PSA case. Just confirming.

Reply:
This lock is used to protect the sign parameters, such as key and sign algorithm is not changed in the signing process. Without the lock, if one task is signing another task change the parameters by calling c_signinit, the signing process will be probably not done successfully. So I think this lock is essential.